lib/facter/settings.rb

```diff
@@ -51,3 +51,16 @@
     File.expand_path(File.join(Puppet.settings['confdir'],'../../mcollective/etc'))
   end
 end
+
+Facter.add('puppet_agent_environment') do
+  setcode do
+    # There has to be a more graceful way to accomplish this, but I cannot seem to find a way
+    # via the existing ruby library in puppet... I think face action :config is worth looking at
+    # but have not been able to make it work, so this will have to do.
+    #
+    # This fact gets the puppet.conf config environment setting independent of the enc override or 
+    # the cli --environment override. The goal is to specify the agent environment to put into the 
+    # puppet.conf file by the msi installer without accidentally overriding it.
+    Facter::Core::Execution.exec('puppet config print --section agent environment');
```

Comment:
@MikaelSmith I am a bit confused by config_by_cli I am not able to find where this is coming from can you provide a blob?

I am also a bit perplexed by only invoking on older versions... Assuming you are talking about `puppet config` and not the blob provided (due to puppet.settings not being available in 3.8 per your previous comment). 

In the interests of code simplicity, perhaps a custom fact is not well placed in this case we need the value exclusively at upgrade and only for windows... Might I recommend executing the `puppet config print --section agent environment` chunk from within the .bat